### DISCUSSION TITLE
Example of using custom `NumberFloatType` ?

### DISCUSSION THREAD

**Discussion by mself** [nlohmann/json#3050] [Q&A] [answered]
I'm trying to follow the docs to use a modified `operator==()` to use a tolerance when comparing floating point JSON values.  I created a custom `MyDouble` class and then tried `using json = nlohmann::basic_json<std::map, std::vector, std::string, bool, std::int64_t, std::uint64_t, MyDouble>;`

But I can't get it to compile (on a Mac) because `json.hpp` uses `std::isfinite()`.  And (on the Mac), the definition of `std::isfinite()` will only accept types where `std::is_arithmetic` is true (i.e only the built-in arithmetic types). It won't compile even if you have a default conversion to `double`.

Does anyone have a working example of how to define a type that can be used for `NumberFloatType` other than `double`?

Here is what I tried:
```
class MyDouble;

using json = nlohmann::basic_json<std::map, std::vector, std::string, bool,
                                  std::int64_t, std::uint64_t, MyDouble>;

/// Class to represent JSON floating point values, but with modified comparison.
class MyDouble {
  double d_;
  static constexpr double epsilon = 1e-3;

 public:
  MyDouble() = default;
  MyDouble(double d) : d_{d} {}
  MyDouble& operator=(const MyDouble&) = default;

  operator double&() { return d_; }
  operator json() { return json(d_); }

  friend bool operator==(MyDouble lhs, MyDouble rhs) {
    return abs(lhs.d_ - rhs.d_) < (epsilon / 2) * (lhs.d_ + rhs.d_);
  }
};
```

Thank you for the awesome library!

**Comment by mself** [comment]
Here is the definition of `isfinite()` on the Mac:

```template <class _A1>
inline _LIBCPP_INLINE_VISIBILITY
typename std::enable_if<
    std::is_arithmetic<_A1>::value && std::numeric_limits<_A1>::has_infinity,
    bool>::type
isfinite(_A1 __lcpp_x) _NOEXCEPT
{
    return __libcpp_isfinite((typename std::__promote<_A1>::type)__lcpp_x);
}

template <class _A1>
inline _LIBCPP_INLINE_VISIBILITY
typename std::enable_if<
    std::is_arithmetic<_A1>::value && !std::numeric_limits<_A1>::has_infinity,
    bool>::type
isfinite(_A1) _NOEXCEPT
{ return true; }
```

Because of this definition, the argument's type must have `std::is_arithmetic` defined, which means it must be a built-in type.

**Reply by mself** [reply]
Perhaps a code fix could be to add a cast to `double` in the code that calls `isfinite()`?

**Reply by nlohmann** [reply]
See https://json.nlohmann.me/features/types/number_handling/#template-number-types

**Comment by mself** [comment]
I can see how to use a different *builtin* numeric type for `NumberFloatType` (i.e. `float`, `double`, or `long double`).

But I'm looking for an example of using a non-builtin type for `NumberFloatType` so that I can define my own `operator==()` function.  I'm not sure that this is currently possible because of the issue with `isfinite()` only working on builtin types.

**Reply by nlohmann** [reply]
I don't think that `std::isfinite` would be the only blocker - also the lexer needs to understand how to parse numbers. To this end, I think the list of possible alternative types for `NumberFloatType` listed in the documentation is exhaustive.

**Comment by mself** [comment]
I now see in the docs that only `float`, `double`, or `long double` are supported for `NumberFloatType`.  But then I find the section on "Alternative comparison functions" confusing, since it seems to imply that there is a way to use your own comparison function without having to provide your own version of the entire JSON equality comparison (where its says "Or you can self-define an operator equal function like this").  There really isn't an "or" -- that is the only way to do it, right?

**Comment by mself** [comment]
So, to rephrase the question, is there a way to just define a custom comparison operator for number values without having to write your own comparison function that handles all of the JSON types (including array and object)?

**Reply by nlohmann** [reply]
If you define your own comparison operator, then you need to make sure it is used whenever you want to compare JSON values. There is no way to change the behavior of `basic_json::operator==`.

**Comment by mself** [comment, marked as answer]
Here is the solution that I ended up with.  It might be helpful to provide an example like this in the section on [Number Comparison](https://json.nlohmann.me/features/types/number_handling/#number-comparison). Thank you for your help!

```cpp
bool equal_within_tolerance(const nlohmann::json& lhs,
                            const nlohmann::json& rhs, double tolerance) {
  if (lhs.is_array() and rhs.is_array()) {
    if (lhs.size() != rhs.size()) return false;
    for (auto i = 0; i < lhs.size(); i++) {
      if (not equal_within_tolerance(lhs[i], rhs[i], tolerance)) return false;
    }
    return true;
  } else if (lhs.is_object() and rhs.is_object()) {
    if (lhs.size() != rhs.size()) return false;
    for (auto& [key, value] : lhs.items()) {
      if (not equal_within_tolerance(value, rhs[key], tolerance)) return false;
    }
    return true;
  } else if (lhs.type() == nlohmann::detail::value_t::number_float and
             rhs.type() == nlohmann::detail::value_t::number_float) {
    return abs(double(lhs) - double(rhs)) <=
           tolerance * (double(lhs) + double(rhs)) / 2;
  } else {
    return lhs == rhs;
  }
}
```
